KEP-3221: Add multiple authorization webhooks KEP (#3376)
kubernetes/enhancements: keps/sig-auth/3221-structured-authorization-configuration/README.md

only allows specifying a single webhook, this would result in cluster
administrators being not able to specify their own. Support for multiple
webhooks would make this possible. Moreover, RBAC rules can't be used here
since RBAC allows to one add permissions but not deny them. We won't be able
to add restrictions on 'non-system' users only for a set of resources using
RBAC; 'non-system' users here refers to anyone who shouldn't be able to
update/delete the protected set of CRDs.

Relevant Discussion Thread is sig-apimachinery: https://groups.google.com/g/kubernetes-sig-api-machinery/c/MBa19WTETMQ

#### Story 2: Preventing unnecessarily nested webhooks

Consider a system administrator who would like to apply a set of validations
to certain requests before handing it off to webhooks defined using frameworks
like Open Policy Agent.

They would have to nested webhooks within the one added to the auth chain to
have the intended effect. This enhancement allows the administrator to configure
this behaviour via a structured API and invoke the additional webhook only when
relevant. This also allows administrators to define `onError` behaviours for

> separate webhooks, leading to more predictable outcomes.

Reviewer: done

`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
respectively.

This diagram shows how the configuration flags map to configuration file keys
for a webhook:
![](https://i.imgur.com/PJSmgOq.jpg)

The `onError` will be a required field which allow users to specify whether
or not request is denied if the webhook errors out or is unreachable. This allows
`SubjectAccessReviews` to be more definitive.

Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
flag is not supplied. While configuring authorization modes using the file config,
the version supported by a webhook has to be mentioned using a required field
`subjectAccessReviewVersion`.

The new structure introduces a configurable timeout that allows administrators to
define a timeout for webhooks in case they are unreachable.

> Note: ABAC will not be supported by the new file format as we discourage the usage of this mode.

The code path for enabling the above will only be triggered if the feature flag will
be enabled until the time the feature flag is removed and configuring authorizer
through a file becomes GA.

> > TODO: More details on the implementation will be charted once the goals, non-goal and proposal is approved.

Author: documented.

### Monitoring

We will add the following 4 metrics:

> 1. `apiserver_authorization_step_invocations_total`

Reviewer: How about if we follow `pod_security_evaluations_total` 
> Number of policy evaluations that occurred, not counting ignored or exempt requests.

We can also exclude `NoOpinion` in the count?

> 1. `apiserver_authorization_step_invocations_total`

Reviewer: Per discussion with @enj, he is ok with 
> We can also exclude NoOpinion in the count

and prefers `apiserver_authorization_decisions_total`

> 1. `apiserver_authorization_step_invocations_total`

Author: I like both the suggested names. Going with `apiserver_authorization_decisions_total` since that seems to be the mutual preference.

Also, dropped `NoOpinion` in the count.

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

> - `mode` {RBAC, Node, Webhook}

Reviewer 2: I was also expecting to see ABAC here (if the cluster enables it). If it's not, let's explain in this KEP why we excluded it.

> - `mode` {RBAC, Node, Webhook}

Author: > will all authorizers show up here? (superuser, abac, etc)

The answer to this is yes. I enumerated the possible values when we planned to deprecate ABAC. Since that is non-concern wrt this KEP, any authorizer should show up in values.

- `decision` {Allow, Deny, NoOpinion}

2. `apiserver_authorization_step_webhook_invocations_total`

This will be incremented on round-trip of an authorization webhook. It will track
total invocation counts across the following labels.

- `server`

> - `code` {2xx, 4xx, 5xx}

Reviewer: How about if we follow the `invalid_key_id_from_status_total` metric:
> Number of times an invalid keyID is returned by the Status RPC call split by error.

such that the error are common error const: https://github.com/kubernetes/kubernetes/blob/8b7c7d783908efb53056fcae121ed18e117a8f2e/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go#L73-L75

> - `code` {2xx, 4xx, 5xx}

Reviewer: Per discussion with @enj, he is ok with ^ and change the metric to `apiserver_authorization_webhook_evaluations_total`

> - `code` {2xx, 4xx, 5xx}

Author: Done.

> - `decision` {Allow, Deny, NoOpinion}

Reviewer: same here, how about we exclude `NoOpinion`
https://github.com/kubernetes/enhancements/pull/3376/files#r1220854280

> - `decision` {Allow, Deny, NoOpinion}

Reviewer: > what metric will we increment if a webhook gets skipped because its matchConditions don't match? not sure if that's better as an additional value in this decision label, or as a separate metric
> 
> cc @logicalhan for thoughts

Similar to `NoOpinion`, if we increase for all requests that skip a webhook due to a matchCondition, the count could be quite large.

> - `decision` {Allow, Deny, NoOpinion}

Reviewer: Per discussion with @enj, he is ok with exclude `NoOpinion` from `decision`. For webhooks that get skipped, that can be done with an audit annotation (e.g. `webhook_skipped`) such that the resource and the user skipping the webhook will also be in the log. TODO: add this to the design section.

- `subject_access_review_version` {v1beta1, v1}

3. `apiserver_authorization_step_webhook_duration_seconds`

This metric will track the average latency

Labels {along with possible values}:
- `server`
- `code` {2xx, 4xx, 5xx}
- `decision` {Allow, Deny, NoOpinion}
- `subject_access_review_version` {v1beta1, v1}

4. `apiserver_authorization_step_webhook_error_total`

This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.

Labels {along with possible values}:

- server
- `code` {4xx, 5xx}
- `decision` {Deny, NoOpinion}
- `subject_access_review_version` {v1beta1, v1}

> ### Test Plan

Reviewer 3: I think we should also benchmark the cost (time/memory/allocs) of some representative CEL matchConditions:
- scoping to a particular namespace
- scoping to a particular group/resource
- excluding a particular set of users
- excluding a particular set of groups

###### What happens if we reenable the feature if it was previously rolled back?

When the feature is re-enabled, `--authorization-config-file` flag should be present.
The behaviour is the same as when the feature is enabled for the first time.

###### Are there any tests for feature enablement/disablement?

> //TODO: How do we ensure functionality in the absence of e2e-framework not having support for feature gates?

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

A rollout can fail when the authorization configuration file being passed doesn't
exist or is invalid.

Already running workloads won't be impacted but cluster users won't be able to
access the control plane if the cluster is single-node.

###### What specific metrics should inform a rollback?

Not applicable.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

TBD.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

None until the feature reaches GA and at a future point in time, `--authorization-*`

> flags other than `--authorization-config-file` are removed.

Reviewer 3: it's unclear whether we would remove the CLI flags or not, that deprecation is not required by the feature addition

The proposed structure is illustrated below:

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: AuthorizationConfiguration
authorizers:
  - type: Webhook
    webhook:
      # Name used to describe the webhook
      # This is explicitly used in monitoring machinery for metrics
      # Note:
      #   - Do exercise caution when setting the value
      #   - If not specified, the default would be set to ""
      #   - If there are multiple webhooks in the authorizer chain,
      #     this field is required
      name: super-important-kube-system-authorizer
      # The duration to cache 'authorized' responses from the webhook
      # authorizer.
      # Same as setting `--authorization-webhook-cache-authorized-ttl` flag
      # Default: 5m0s
      authorizedTTL: 30s
      # The duration to cache 'authorized' responses from the webhook
      # authorizer.
      # Same as setting `--authorization-webhook-cache-unauthorized-ttl` flag
      # Default: 30s
      unauthorizedTTL: 30s
      # Timeout for the webhook request
      # Maximum allowed is 30s.
      # Required, with no default.
      timeout: 3s
      # The API version of the authorization.k8s.io SubjectAccessReview to

>       # send to and expect from the # webhook.

Reviewer 3: ```suggestion
      # send to and expect from the webhook.
```

      # Same as setting `--authorization-webhook-version` flag
      # Required, with no default
      subjectAccessReviewVersion: v1
      # Controls the authorization decision when a webhook request fails to

>       # complete or returns a malformed response.

Reviewer 3: ```suggestion
      # complete or returns a malformed response or errors evaluating matchConditions.
```

      # Valid values:
      #   - NoOpinion: continue to subsequent authorizers to see if one of
      #     them allows the request
      #   - Deny: reject the request without consulting subsequent authorizers
      # Default: NoOpinion
      failurePolicy: Deny
      connectionInfo:
        # Controls how the webhook should communicate with the server.
        # Valid values:
        # - KubeConfig: use the file specified in kubeConfigFile to locate the
        #   server.
        # - InClusterConfig: use the in-cluster configuration to call the
        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
        #   allowed for kube-apiserver.
        type: KubeConfig
        # Path to KubeConfigFile for connection info
        kubeConfigFile: /kube-system-authz-webhook.yaml

>       matchConditions:

Reviewer 3: add docs for the `matchConditions`:

```
	# matchConditions is a list of conditions that must be met for a request to be sent to this
	# webhook. An empty list of matchConditions matches all requests.
	# There are a maximum of 64 match conditions allowed.
	#
	# The exact matching logic is (in order):
	#   1. If ANY matchCondition evaluates to FALSE, the webhook is skipped.
	#   2. If ALL matchConditions evaluate to TRUE, the webhook is called.
	#   3. If any matchCondition evaluates to an error (but none are FALSE):
	#      - If failurePolicy=Deny, reject the request
	#      - If failurePolicy=NoOpinion, the error is ignored and the webhook is skipped
```

>       - expression: |

Reviewer 3: add docs for the expression field:

```
	# expression represents the expression which will be evaluated by CEL. Must evaluate to bool.
	# CEL expressions have access to the contents of the SubjectAccessReview
  # in the version specified by subjectAccessReviewVersion in the request variable.
	# Documentation on CEL: https://kubernetes.io/docs/reference/using-api/cel/
```

>         request.resourceAttributes.namespace == 'kube-system'

Reviewer 3: @ritazh - as a follow-up, check how this expression would behave for a non-resource request (e.g. to `/version`) where resourceAttributes is null... if this would error, we should improve this example to put a `has(request.resourceAttributes)` match condition first, etc

Author: I have added the conditional either way.

invocations.
- Enable user to define the policy when a webhook can't be reached due to
network errors etc.
- Enable hot reload of the configuration.

### Non-Goals

- Replace path to kubeconfig file for webhooks by allowing auth configuration
within the new file structure
- Allow ordering of super user authorizer (always allow user group;
`system:masters`) so that it can be moved from the beginning of the auth chain.
- Deciding whether ABAC should be continued to be used by the users.

> Note: The above may be revisited in future alpha iterations of this KEP or as a separate KEP (for e.g., the superuser authorizer)

## Proposal

Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain. See [Design Details](#design-details) for more details.

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: AuthorizationConfiguration
authorizers:
  - type: Webhook
    webhook:
      unauthorizedTTL: 30s
      timeout: 3s
      subjectAccessReviewVersion: v1
      failurePolicy: Deny
      connectionInfo:
        type: KubeConfig
        kubeConfigFile: /kube-system-authz-webhook.yaml

>       matchConditions:

Reviewer 3: let's make the example a little cleaner for people to copy:

```yaml
matchConditions:
# only send resource requests to the webhook
- expression: has(request.resourceAttributes)
# only intercept requests to kube-system
- expression: request.resourceAttributes.namespace == 'kube-system'
# don't intercept requests from kube-system service accounts
- expression: !('system:serviceaccounts:kube-system' in request.user.groups)
```

      - expression: |
        has(request.resourceAttributes) && (request.resourceAttributes.namespace == 'kube-system')
      - expression: |
        !('system:serviceaccounts:kube-system' in request.user.groups)
  - type: Node
  - type: RBAC
  - type: Webhook
    webhook:
      authorizedTTL: 5m
      unauthorizedTTL: 30s

>       timeout: 3s

Reviewer 2: I had a concern that's got lost - https://github.com/kubernetes/enhancements/pull/3376#discussion_r1204487517

I don't have any concern about this going to alpha. Could we track it as a detail to make sure we consider before this moves to GA?

Reviewer 3: there are only two examples of using field names with units in file-based config

```
NodeLeaseDurationSeconds int32 `json:"nodeLeaseDurationSeconds,omitempty"`
ShutdownGracePeriodSeconds int64 `json:"shutdownGracePeriodSeconds"`
PodInitialBackoffSeconds *int64 `json:"podInitialBackoffSeconds,omitempty"`
PodMaxBackoffSeconds *int64 `json:"podMaxBackoffSeconds,omitempty"`
```

there are 35+ examples of using metav1.Duration in config files, including for things like kubelet webhook authentication:
```
CacheTTL metav1.Duration `json:"cacheTTL,omitempty"`
```
credentialprovider cache duration:
```
CacheDuration *metav1.Duration `json:"cacheDuration,omitempty"`
```

I think the time.Duration type is more usable and more consistent. I don't think we'll change this.

Reviewer 2: > validation around the timeout so that it cannot exceed 30 seconds

Ah, nice.

cluster bootstrapper handle such scenarios.
- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization
to be configured using a flag (`--authorization-config-file`) which accepts a
path to a file on the disk. Setting both `--authorization-config-file` and
configuring an authorization webhook will not be allowed. If the user does that,
there will be an error and API Server would exit right away.

The configuration would be validated at startup and the API server will fail to
start if the configuration is invalid.

The API server will periodically reload the configuration. If it changes, the
new configuration will be used for the Authorizer chain. If the new configuration
is invalid, the last known valid configuration will be used. Logging and metrics
would be used to signal success/failure of a config reload so that cluster admins

> can have observability over this process.

Reviewer 3: Add a note that the reload must not add or remove Node or RBAC authorizers... whatever the apiserver starts with is what any reloaded file must have. They can be reordered, but not added/removed.

Reviewer: If the webhook is not available, then doesn't that make the new configuration invalid, which ensures the last known valid configuration will be used. Logging and metrics
can be used to signal this type of failure of a config reload so that cluster admins
can have observability over this process.

Reviewer: Per discussion with @enj,
We need a time-based loop inside the apiserver to check if the file changed to trigger an update. If healthcheck on the wehobook failed, we will use the last known good config. The time-based loop will try later and when webhook health comes back, we will use the new config.

Author: The time-based loop alleviates my concerns with retrying non-existing webhook URLs. This way we don't need to specify a retry count with backoff mechanisms.

Adding the same to the Design Details.

to a `request` variable containing a `SubjectAccessReview` object in the version specified
by `subjectAccessReviewVersion`.

The code path for enabling the above will only be triggered if the feature flag will
be enabled until the time the feature flag is removed and configuring authorizer
through a file becomes GA.

### Monitoring

We will add the following 4 metrics:

1. `apiserver_authorization_step_invocations_total`

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

Labels {along with possible values}:
- `mode` {<authorizer_name>}
- `decision` {Allow, Deny, NoOpinion}

**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.
If there is only one webhook, there would be no `_<name>` suffix.

2. `apiserver_authorization_step_webhook_invocations_total`

This will be incremented on round-trip of an authorization webhook. It will track
total invocation counts across the following labels.

- `name`
- `code` {2xx, 4xx, 5xx}
- `decision` {Allow, Deny, NoOpinion}

3. `apiserver_authorization_step_webhook_duration_total_seconds`

> This metric will track the total round trip time of the requests to the webhook.

Reviewer: Should this be Histogram instead of count so we can see duration/latency across timespan?

Author: I like the idea! This gives more power to the users. Thanks!

Author: Also, followed the labels as `apiserver_authorization_webhook_evaluations_total` plus added `errAuthzWebhookOKCode = "ok"` as I feel it's a valid use case to understand the average latency of an webhook request across time even if it is a successful one.
```
- `code` {"incomplete_request", "bad_response", "ok"}
```

Labels {along with possible values}:
- `name`
- `code` {2xx, 4xx, 5xx}
- `decision` {Allow, Deny, NoOpinion}

> 4. `apiserver_authorization_step_webhook_error_total`

Reviewer: Since we have `apiserver_authorization_step_webhook_invocations_total` above which captures count for a webhook, code, and decision, this seems redudant.

Reviewer: `apiserver_authorization_webhook_evaluations_fail_open_total` record only when code != `errAuthzWebhookOKCode` and decision on error is not set to`deny`

This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.

Labels {along with possible values}:

- `name`
- `code` {4xx, 5xx}
- `decision` {Deny, NoOpinion}

> 5. `apiserver_authorization_step_configuration_reload_failure_count`

Reviewer: Shouldn't this be Gauge instead of count so we can see last timestamp for current state?

How about `apiserver_authorization_step_configuration_reload_last_timestamp_seconds`
```
The last time in seconds when an authorization reload was performed.
- `name`
- `status` {success, fail}
```

Author: I agree. A gauge is more appropriate.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

TBD.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

None.

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

> //TODO: To be elaborated more during Beta graduation.

###### How can an operator determine if the feature is in use by workloads?

The cluster administrators can check the flags passed to the `kube-apiserver` if
they have access to the control plane nodes. If the `--authorization-config-file`
is set to a valid authorization configuration file, the feature is being used.
Or, they can look at the metrics exposed by `kube-apiserver`.

###### How can someone using this feature know that it is working for their instance?

- [x] Other
  - Details: They can look at the metrics if `apiserver_authorization_step_invocations_total` is increasing.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> The amount of errors denoted by `apiserver_authorization_step_webhook_error_total`

Reviewer: ```suggestion
The total number of `apiserver_authorization_step_webhook_error_total` metric with failure code
```

Author: Note for myself: Revisiting this after resolving other comments around the metric.

Let's say if an API request is being made by an user in a specific group, the
webhook request can be skipped.

### Risks and Mitigations

- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
each are configured. This may create inconsistencies. Mitigation is to have the
cluster bootstrapper handle such scenarios.
- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization
to be configured using a flag (`--authorization-config-file`) which accepts a
path to a file on the disk. Setting both `--authorization-config-file` and
configuring an authorization webhook using the `--authorization-webhook-*` 
command line flags will not be allowed. If the user does that,
there will be an error and API Server would exit right away.

The configuration would be validated at startup and the API server will fail to
start if the configuration is invalid.

> The API server will periodically reload the configuration. If it changes, the

Author: Added in the section that webhooks would be validated too.

`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
respectively.

Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
flag is not supplied. While configuring authorization modes using the file config,
the version supported by a webhook has to be mentioned using a required field
`subjectAccessReviewVersion`.

The user can define a CEL expression to determine whether a request needs to dispatched
to the authz webhook for which the expression has been defined. The user would have access
to a `request` variable containing a `SubjectAccessReview` object in the version specified
by `subjectAccessReviewVersion`.

The code path for enabling the above will only be triggered if the feature flag is enabled until 
the feature flag is removed and this feature graduates to GA.

### Monitoring

We will add the following 4 metrics:

1. `apiserver_authorization_step_invocations_total`

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

Labels {along with possible values}:

> - `mode` {<authorizer_name>}

Reviewer: > - `mode` {<authorizer_name>} when authorizer is a webhook, prepend `webhook_`

Author: There is a note a couple of lines below which specifies this information. I think that conveys the same information in addition to specifying what happens when there is only one Webhook and no name is specified.

Reviewer: Can you pls explicitly add `when authorizer is a webhook, prepend webhook_` here so we do not forget it during implementation.

This metric will track the total round trip time of the requests to the webhook.

Labels {along with possible values}:
- `name`
- `code` {2xx, 4xx, 5xx}
- `decision` {Allow, Deny, NoOpinion}

4. `apiserver_authorization_step_webhook_error_total`

This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.

Labels {along with possible values}:

- `name`
- `code` {4xx, 5xx}
- `decision` {Deny, NoOpinion}

5. `apiserver_authorization_step_configuration_reload_failure_count`

This metric would track the number of times the configuration has been reloaded.

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Unit tests would be added along with any new code introduced.

Existing test coverage:
- cmd/kube-apiserver/app/options: 65
- staging/src/k8s.io/apiserver/pkg/authorization: NA
- staging/src/k8s.io/apiserver/apis/config/v1: 54.5

> Note: We will add more packages to the list during implementation.

**Benchmarks**

> We should benchmark the cost of some common CEL expressions inside

Reviewer: @jpbetz we should do this for CEL in general since this is not specific to this KEP only. WDYT?

Reviewer: xref: https://github.com/google/cel-go/issues/729

Reviewer 4: Ensuring that CEL performance doesn't regress is something I'd like to see owned primarily by cel-go. I suspect there are already things in place. I added a comment on https://github.com/google/cel-go/issues/729 to see if anything can be shared.

That still leaves room for benchmarking on the k8s side. The [CEL Cost model](https://docs.google.com/spreadsheets/d/123G4mLDTWjpCCbfop7HL3NkDvIBLaPu6YqV8ro3Isnw/edit#gid=1510432316) data might be useful in some contexts.

I think it is reasonable for features like proposed by this KEP to add some go mincor benchmarks a cases you care about. This is a relatively low effort thing to do. Typically when people see a PR that they're worried will impact a micro benchmark they ask the PR author to paste before/after output into the PR description.

We do have line items to better track performance before we take features to GA.  I think we'll spend some time in 1.29 on it. It might be that we rely more on metrics than on micro benchmarks?  I'm not sure yet.

each are configured. This may create inconsistencies. Mitigation is to have the
cluster bootstrapper handle such scenarios.
- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization
to be configured using a flag (`--authorization-config-file`) which accepts a
path to a file on the disk. Setting both `--authorization-config-file` and
configuring an authorization webhook using the `--authorization-webhook-*`
command line flags will not be allowed. If the user does that,
there will be an error and API Server would exit right away.

The configuration would be validated at startup and the API server will fail to
start if the configuration is invalid.

The API server will periodically reload the configuration. If it changes, the
new configuration will be used for the Authorizer chain. If the new configuration
is invalid, the last known valid configuration will be used. The reloader will also

> check if the webhook exists, thereby preventing any typo/misconfiguration with the

Reviewer: ```suggestion
check if the webhook is healthy, thereby preventing any typo/misconfiguration with the
```

> Webhook resulting in bad Authorizer config. Logging and metrics would be used to

Reviewer: ```suggestion
Webhook resulting in bad Authorizer config. If healthcheck on the wehobook failed, the last known good config will be used. The time-based loop will try later and when webhook health comes back, the new config will be used. Logging and metrics would be used to
```

Author: I have rewritten the whole block to simplify it.

TODO(@palnabarun): Add a flow diagram for the check. This would help to understand visually.

flag is not supplied. While configuring authorization modes using the file config,
the version supported by a webhook has to be mentioned using a required field
`subjectAccessReviewVersion`.

The user can define a CEL expression to determine whether a request needs to dispatched
to the authz webhook for which the expression has been defined. The user would have access
to a `request` variable containing a `SubjectAccessReview` object in the version specified
by `subjectAccessReviewVersion`.

When no matchConditions are satisfied for a request, the webhook would be skipped. In such
situations, the decision is logged in the audit log with the `webhook_skipped` annotation.
Benefit of this is that resource and user info will also be logged.

The code path for enabling the above will only be triggered if the feature flag is enabled until
the feature flag is removed and this feature graduates to GA.

### Monitoring

We will add the following 4 metrics:

1. `apiserver_authorization_decisions_total`

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

Labels {along with possible values}:

> - `mode` {<authorizer_name>}

Reviewer: ```suggestion
- `mode` {<authorizer_name>} # when authorizer is a webhook, prepend webhook_
```

- `decision` {Allow, Deny}

**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.
If there is only one webhook, there would be no `_<name>` suffix.

2. `apiserver_authorization_webhook_evaluations_total`

This will be incremented on round-trip of an authorization webhook. It will track
total invocation counts across the following labels.

- `name`
- `code` {"incomplete_request", "bad_response"}

3. `apiserver_authorization_webhook_evaluations_duration_seconds`

This is a Histogram metric that will track the total round trip time of the requests to the webhook.

Labels {along with possible values}:
- `name`
- `code` {"incomplete_request", "bad_response", "ok"}

> 4. `apiserver_authorization_step_webhook_error_total`

Reviewer: ```suggestion
4. `apiserver_authorization_webhook_evaluations_fail_open_total`
```

Reviewer: we have `apiserver_admission_webhook_fail_open_count` metric today

> This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.

Reviewer: ```suggestion
This metric will be incremented when a webhook returns `code != errAuthzWebhookOKCode` and decision on error is not set to `deny`.
```

Labels {along with possible values}:

- `name`

> - `code` {4xx, 5xx}

Reviewer: ```suggestion
- `code` {"incomplete_request", "bad_response"}
```

> - `decision` {Deny, NoOpinion}

Reviewer: ```suggestion
- `decision` {Deny}
```

Author: Do we need the `decision` label if the only value is going to be `NoOpinion`?

Reviewer: Good point, please remove it.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

TBD.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

None.

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

> //TODO: To be elaborated more during Beta graduation.

###### How can an operator determine if the feature is in use by workloads?

The cluster administrators can check the flags passed to the `kube-apiserver` if
they have access to the control plane nodes. If the `--authorization-config-file`
is set to a valid authorization configuration file, the feature is being used.
Or, they can look at the metrics exposed by `kube-apiserver`.

###### How can someone using this feature know that it is working for their instance?

- [x] Other
  - Details: They can look at the metrics if `apiserver_authorization_decisions_total` is increasing.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> The total number of `apiserver_authorization_step_webhook_error_total` metric with failure code

Reviewer: ```suggestion
The total number of `apiserver_authorization_webhook_evaluations_fail_open_total` metric with failure code
```

cluster bootstrapper handle such scenarios.
- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization
to be configured using a flag (`--authorization-config-file`) which accepts a
path to a file on the disk. Setting both `--authorization-config-file` and
configuring an authorization webhook using the `--authorization-webhook-*`
command line flags will not be allowed. If the user does that,
there will be an error and API Server would exit right away.

The configuration would be validated at startup and the API server will fail to
start if the configuration is invalid.

The API server will periodically reload the configuration at a specific time
interval. If it changes, the new configuration will be used for the Authorizer
chain. The reloader will also check if the webhook is healthy, thereby
preventing any typo/misconfiguration with the Webhook resulting in bad

> Authorizer config. If healthcheck on the webhook failed, the last known good

Author: @ritazh I noticed that the specific terms `healthy` and `healthcheck`. I was curious if there are the checks we can do to determine health other than checking for existence of the webhook, dispatching a dummy request and getting the response?

I want to specify explicitly in the KEP what kind of checks will be performed.

Reviewer: `checking for existence of the webhook` isn't as clear to me what specifically we need to check, whereas `using the webhook healthcheck` explicitly tells us what we need to check to ensure the webhook is functioning as intended.

grants access to the user.

#### Story 5: Varying defaults across versions of the API

If the default values for configuration evolve over time, affected users might
have to override these values in case they are affected by supplying the flags
with updated defaults. A configuration format like the one proposed can be
versioned over time with changing defaults, mitigating this risk by only
affecting users who knowingly bump the version.

#### Story 6: Conditionally filtering requests to webhooks

Let's say if an API request is being made by an user in a specific group, the
webhook request can be skipped.

### Risks and Mitigations

- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
each are configured. This may create inconsistencies. Mitigation is to have the
cluster bootstrapper handle such scenarios.
- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization

> to be configured using a flag (`--authorization-config-file`) which accepts a

Reviewer 5: ```suggestion
to be configured using a flag (`--authorization-config`) which accepts a
```

Reviewer 5: This needs to be symmetric with #3332 which uses `--authentication-config`.  We are not super consistent in this space because I see both `--encryption-provider-config` and `--admission-control-config-file` as current flags to the API server.

Reviewer: done

preventing any typo/misconfiguration with the Webhook resulting in bad
Authorizer config. If healthcheck on the webhook failed, the last known good
config will be used. In the next iteration of reload, if webhook is found to be
healthy, the new config will be used. Logging and metrics would be used to
signal success/failure of a config reload so that cluster admins can have
observability over this process. Reload must not add or remove Node or RBAC
authorizers. They can be reordered, but cannot be added or removed.

The proposed structure is illustrated below:

> The sample configuration describes all the fields, their defaults and possible
values.

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: AuthorizationConfiguration
authorizers:
  - type: Webhook
    webhook:
      # Name used to describe the webhook
      # This is explicitly used in monitoring machinery for metrics
      # Note:
      #   - Do exercise caution when setting the value
      #   - If not specified, the default would be set to ""
      #   - If there are multiple webhooks in the authorizer chain,
      #     this field is required

>       name: super-important-kube-system-authorizer

Reviewer: This `name` field is new, for existing webhooks configured via the command line flag, what value would we use?

>       name: super-important-kube-system-authorizer

Author: @enj what are your opinions on the above?

>       name: super-important-kube-system-authorizer

Reviewer 5: `my_fancy_webhook`?  The value doesn't really matter because it is meant as a human friendly name.  But I do not really see value is allowing it to be optional.

      # authorizer.
      # Same as setting `--authorization-webhook-cache-authorized-ttl` flag
      # Default: 5m0s
      authorizedTTL: 30s
      # The duration to cache 'authorized' responses from the webhook
      # authorizer.
      # Same as setting `--authorization-webhook-cache-unauthorized-ttl` flag
      # Default: 30s
      unauthorizedTTL: 30s
      # Timeout for the webhook request
      # Maximum allowed is 30s.
      # Required, with no default.
      timeout: 3s
      # The API version of the authorization.k8s.io SubjectAccessReview to
      # send to and expect from the webhook.
      # Same as setting `--authorization-webhook-version` flag
      # Required, with no default
      # Valid values: v1beta1, v1
      subjectAccessReviewVersion: v1
      # Controls the authorization decision when a webhook request fails to
      # complete or returns a malformed response or errors evaluating
      # matchConditions.
      # Valid values:
      #   - NoOpinion: continue to subsequent authorizers to see if one of
      #     them allows the request
      #   - Deny: reject the request without consulting subsequent authorizers

>       # Default: NoOpinion

Author: I am different on this. Having `failurePolicy` Required with no default also sounds good to me.

Right 

> Re: SubjectAccessReview are you saying we should make it required and not default to v1beta1?

I think Mo is saying that we need to be consistent with the field defaulting behaviour.

>       # Default: NoOpinion

Author: Changed to Required, with no default.

      failurePolicy: Deny
      connectionInfo:
        # Controls how the webhook should communicate with the server.
        # Valid values:
        # - KubeConfig: use the file specified in kubeConfigFile to locate the
        #   server.
        # - InClusterConfig: use the in-cluster configuration to call the
        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
        #   allowed for kube-apiserver.
        type: KubeConfig
        # Path to KubeConfigFile for connection info
        # Required, if connectionInfo.Type is KubeConfig
        kubeConfigFile: /kube-system-authz-webhook.yaml
        # matchConditions is a list of conditions that must be met for a request to be sent to this
        # webhook. An empty list of matchConditions matches all requests.
        # There are a maximum of 64 match conditions allowed.
        #
        # The exact matching logic is (in order):
        #   1. If ANY matchCondition evaluates to FALSE, the webhook is skipped.
        #   2. If ALL matchConditions evaluate to TRUE, the webhook is called.

>         #   3. If any matchCondition evaluates to an error (but none are FALSE):

Reviewer 5: This is a bit difficult to parse, I feel like it would be easier to explain how each item in the list is handled.

        #      - If failurePolicy=Deny, reject the request
        #      - If failurePolicy=NoOpinion, the error is ignored and the webhook is skipped
      matchConditions:
      # expression represents the expression which will be evaluated by CEL. Must evaluate to bool.
      # CEL expressions have access to the contents of the SubjectAccessReview

>       # in the version specified by subjectAccessReviewVersion in the request variable.

Reviewer 5: Do we really want to tie the CEL expression input to the version of the webhook?

Author: We need to support both versions for the time being. I don't see a path forward without tying to the version unless we internally convert between the versions, which would make it more complicated and less transparent to users.

Reviewer 5: I think it would be better to convert internally to v1 always because it keeps all of our CEL examples consistent and valid regardless of how the webhook works.  Doubling the CEL surface area would be a mistake.

Author: Doesn't it look odd in terms of the config if we allow both versions for the webhook request, but then have CEL filters apply to only v1 after internal conversions?

Reviewer 5: Think about this way: does it really provide value to the end user to have to learn _two_ slightly different ways to write CEL filters?  The user configuring the API server may not be the webhook author.  IMO we are conflating two unrelated things here - one is about how the API server makes decisions, and the other is about how we ask an external webhook how to make decisions.  These are not the same, and CEL is net new and has no reason to support a beta format.

Author: > The user configuring the API server may not be the webhook author.

Correct.

> These are not the same, and CEL is net new and has no reason to support a beta format.

Makes sense.

      # Documentation on CEL: https://kubernetes.io/docs/reference/using-api/cel/
      #
      # only send resource requests to the webhook
      - expression: has(request.resourceAttributes)
      # only intercept requests to kube-system
      - expression: request.resourceAttributes.namespace == 'kube-system'
      # don't intercept requests from kube-system service accounts
      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
  - type: Node
  - type: RBAC
  - type: Webhook
    webhook:
      name: in-cluster-authorizer
      authorizedTTL: 5m
      unauthorizedTTL: 30s
      timeout: 3s
      subjectAccessReviewVersion: v1
      failurePolicy: NoOpinion
      connectionInfo:
        type: InClusterConfig
```

Validation will allow multiple authorizers of type "Webhook" to be added to the

> config, but one authorizer each for other types. The ordering of this chain will

Reviewer 5: ```suggestion
config, but at most one authorizer each for other types. The ordering of this chain will
```

be decided by the order specified in the file.

The keys `kubeConfigFile`, `authorizedTTL`, `unauthorizedTTL` and
`subjectAccessReviewVersion` accept values corresponding to flags
`--authorization-webhook-config-file`, `--authorization-webhook-cache-authorized-ttl`,
`--authorization-webhook-cache-unauthorized-ttl` and `--authorization-webhook-version`
respectively.

Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
flag is not supplied. While configuring authorization modes using the file config,
the version supported by a webhook has to be mentioned using a required field
`subjectAccessReviewVersion`.

The user can define a CEL expression to determine whether a request needs to dispatched
to the authz webhook for which the expression has been defined. The user would have access
to a `request` variable containing a `SubjectAccessReview` object in the version specified
by `subjectAccessReviewVersion`.

When no matchConditions are satisfied for a request, the webhook would be skipped. In such

> situations, the decision is logged in the audit log with the `webhook_skipped` annotation.

Reviewer 5: The annotation needs to be namespaced to the `k8s.io` domain.

Benefit of this is that resource and user info will also be logged.

The code path for enabling the above will only be triggered if the feature flag is enabled until

> the feature flag is removed and this feature graduates to GA.

Reviewer 5: I do not understand this sentence.

### Monitoring

We will add the following 4 metrics:

1. `apiserver_authorization_decisions_total`

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

Labels {along with possible values}:
- `mode` {<authorizer_name>} # when authorizer is a webhook, prepend `webhook_`
- `decision` {Allow, Deny}

**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.

> If there is only one webhook, there would be no `_<name>` suffix.

Reviewer 5: We should document that the webhook name will be "exposed" via the metrics endpoint.

decision on error is not set to `deny`.

Labels {along with possible values}:

- `name`
- `code` {"incomplete_request", "bad_response"}

5. `apiserver_authorization_configuration_reload_last_timestamp_seconds`

This Gauge metric will record last time in seconds when an authorization reload was performed.
- `status` {success, fail}

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

Unit tests would be added along with any new code introduced.

Existing test coverage:
- cmd/kube-apiserver/app/options: 65
- staging/src/k8s.io/apiserver/pkg/authorization: NA
- staging/src/k8s.io/apiserver/apis/config/v1: 54.5

> Note: We will add more packages to the list during implementation.

**Benchmarks**

We should benchmark the cost of some common CEL expressions inside
`matchConditions`. Some examples being:
- scoping to a particular namespace/group/resource
- excluding a particular set of users/groups

> ##### Integration tests

Reviewer 5: Automatic reload as well?

- Unit tests and Integration tests to be written

#### Future Alpha versions

- Revisit on the items mentioned in Non-Goals and see if any needs to be implemented

#### Beta

- Address user reviews and iterate (if needed, keep in Alpha until changes stabilize)
- Feature flag will be turned on by default

#### GA

- Feature flag removed
- Existing command line flags will be marked as deprecated and config file will take
precedence over the old flags

#### GA + 3 cycles

- Remove the existing command line flags

### Upgrade / Downgrade Strategy

While the feature is in Alpha, there is no change if cluster administrators want to
keep on using command line flags.

When the feature goes to Beta/GA or the cluster administrators want to configure
authorizers using a config file, they need to make sure the config file exists before
upgrading the cluster. Similarly when downgrading clusters, they would need to add
the flags back to their bootstrap mechanism.

### Version Skew Strategy

Not applicable.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate

>   - Feature gate name: `AuthorizationConfigFromFile`

Reviewer 5: ```suggestion
  - Feature gate name: `StructuredAuthorizationConfig`
```

Reviewer: IMO it's better to keep the feature gate as `AuthorizationConfigFromFile` to be consistent with the flag `--authorization-config-file`.

>   - Components depending on the feature gate:

Reviewer 5: I forget how we feature gate things that impact aggregated API servers?

Reviewer 6: > I forget how we feature gate things that impact aggregated API servers?

The gate lives in k8s.io/apiserver.

Author: Good to resolve this now?

###### Does enabling the feature change any default behavior?

No.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, the feature can be disabled once enabled.

###### What happens if we reenable the feature if it was previously rolled back?

When the feature is re-enabled, `--authorization-config-file` flag should be present.
The behaviour is the same as when the feature is enabled for the first time.

###### Are there any tests for feature enablement/disablement?

We will have extensive unit tests during feature implementation. There would be tests
for the Authorizer chain in both the old and new configuration scenarios.

We will add integration tests to validate the enablement/disablement flow.
  - When the feature is disabled, only the existing command line flag `--authorization-webhook-*`
based mode is allowed.
  - When the feature is enable, setting both `--authorization-config-file` and
configuring an authorization webhook using the `--authorization-webhook-*`
command line flags should return an error.

### Rollout, Upgrade and Rollback Planning

> Note: This section is required when targeting Beta to a release.

> ###### How can a rollout or rollback fail? Can it impact already running workloads?

Reviewer 5: On rollback, the admin needs to set the old flags right?

A rollout can fail when the authorization configuration file being passed doesn't
exist or is invalid.

Already running workloads won't be impacted but cluster users won't be able to
access the control plane if the cluster is single-node.

###### What specific metrics should inform a rollback?

Not applicable.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

TBD.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

> None.

Reviewer 5: Eventually the old flags go away right?

Reviewer: Good catch. updated.

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

> Note: To be elaborated more during Beta graduation since this section
must be completed when targeting beta to a release.

###### How can an operator determine if the feature is in use by workloads?

The cluster administrators can check the flags passed to the `kube-apiserver` if
they have access to the control plane nodes. If the `--authorization-config-file`
is set to a valid authorization configuration file, the feature is being used.
Or, they can look at the metrics exposed by `kube-apiserver`.

###### How can someone using this feature know that it is working for their instance?

- [x] Other

>   - Details: They can look at the metrics if `apiserver_authorization_decisions_total` is increasing.

Reviewer 5: This would increment even if this feature was not in use (this is a pre-existing metric right?).

Reviewer: This is a new metric similar to all other metrics mentioned in this KEP.

Reviewer: I see the confusion now. I have updated to:
> Details: Since this feature introduced the `name` field to the webhook authorizer,
  users can first specify a value in the `name` field of the AuthorizationConfiguration.
  Then look at the `apiserver_authorization_webhook_evaluations_total` metrics to ensure the
  count for the named webhook authorizer is increasing.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

The total number of `apiserver_authorization_webhook_evaluations_fail_open_total` metric with failure code
is within reasonable limits. A rising value indicates issues with either the
authorizer chain or the webhook itself.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- [x] Metrics
  - Metric name: `apiserver_authorization_decisions_total`

>   - Components exposing the metric: kube-apiserver

Reviewer 5: Same question as above.

**Note**: This is a result of the intended feature.
If multiple webhooks are defined and one or more of them are unreachable, the
request latency would get a hit but this is upto the configuration made by the
user. The feature implementation itself doesn't introduce any change to the
existing SLIs/SLOs.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

<!--
Things to keep in mind include: additional in-memory state, additional
non-trivial computations, excessive access to disks (including increased log
volume), significant amount of data sent and/or received over network, etc.
This through this both in small and large cases, again with respect to the
[supported limits].

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->

No. There would be very minimal addition to the memory used by the API Server and
number of log entries written to the disk. During the Alpha implementation,
the small impact will be measured and rationalized to keep the addition
minimal. The addition would be well within the scalability limits and

> thresholds.

Reviewer 5: I would argue that this feature could make common use cases of a single webhook even faster by introducing the CEL expression filtering to limit the number of webhook calls.

Author: Added a note.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Today, `kube-apiserver` allows one to configure the authorization chain using
a set of command line flags of the format `--authorization-*`. And cluster
administrators can only enable one webhook as a part of the authorization
chain using the `--authorization-modes` flag. This limits admins from creating
authorization chains with multiple webhooks that validate requests in a certain
order. This proposal makes the case for a more structured config to define the
authorization chain for `kube-apiserver` and allowing multiple webhooks with
well-defined parameters and enabling fine grained control like explicit Deny
authorizer.

## Motivation

Cluster administrators should be able to specify more than one authorization
webhook in the API Server handler chain. They also need to be able to
declaratively configure the authorizer chain using a configuration file. It

> should also be easy to say when to Deny requests, for example, when a webhook

Reviewer 7: ```suggestion
should also be easy to say when to deny requests, for example, when a webhook
```

      # don't intercept requests from kube-system service accounts
      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
  - type: Node
  - type: RBAC
  - type: Webhook
    webhook:
      unauthorizedTTL: 30s
      timeout: 3s
      subjectAccessReviewVersion: v1
      failurePolicy: Deny
      connectionInfo:
        type: KubeConfig
        kubeConfigFile: /opa-kube-system-authz-webhook.yaml
      matchConditions:
      # only send resource requests to the webhook
      - expression: has(request.resourceAttributes)
      # only intercept requests to kube-system
      - expression: request.resourceAttributes.namespace == 'kube-system'
      # don't intercept requests from kube-system service accounts
      - expression: !('system:serviceaccounts:kube-system' in request.user.groups)
```

#### Story 3: Denying requests on certain scenarios

The authorizer chain should be powerful enough to deny anyone making a request
if certain conditions are satisfied, except for the `system:masters` user group.

#### Story 4: Controlling access of a privileged RBAC role

Certain users associated with a privileged role might need to have their access

> scoped to certain namespaces. Having ordered authorisation modes allows the

Reviewer 7: ```suggestion
scoped to certain namespaces. Having ordered authorization modes allows the
```

administrator to add a webhook restricting certain user tokens before RBAC
grants access to the user.

#### Story 5: Varying defaults across versions of the API

If the default values for configuration evolve over time, affected users might
have to override these values in case they are affected by supplying the flags
with updated defaults. A configuration format like the one proposed can be
versioned over time with changing defaults, mitigating this risk by only
affecting users who knowingly bump the version.

#### Story 6: Conditionally filtering requests to webhooks

Let's say if an API request is being made by an user in a specific group, the
webhook request can be skipped.

### Risks and Mitigations

- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
each are configured. This may create inconsistencies. Mitigation is to have the

> cluster bootstrapper handle such scenarios.

Reviewer: changed to `cluster administrator`

> cluster bootstrapper handle such scenarios.

Author: Here I had meant bootstrappers like kubeadm can ensure that there are no such inconsistencies. `administrator` also sounds correct here.

- In case an administrator enables this feature and the webhook kubeconfig file
is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
will not be able to start. This can be mitigated by fixing the malformed values.

## Design Details

We would like to introduce a structured file format which allows authorization
to be configured using a flag (`--authorization-config`) which accepts a
path to a file on the disk. Setting both `--authorization-config` and
configuring an authorization webhook using the `--authorization-webhook-*`
command line flags will not be allowed. If the user does that,

> there will be an error and API Server would exit right away.

Reviewer 5: I would add something like this to the KEP (this minimizes that time that any difference in authorization exists between servers):

- Carefully translate existing CLI flag config to the structured config, keep it exactly 1:1
- Change to structured config on each API server and restart one at a time (there should be no change in behavior)
- In parallel across all API servers, update the config files on disk with the final desired structured config (servers will have skewed config on the order of minutes)

> there will be an error and API Server would exit right away.

Author: Added a note.

The user can define a CEL expression to determine whether a request needs to dispatched
to the authz webhook for which the expression has been defined. The user would have access
to a `request` variable containing a `SubjectAccessReview` object in the version specified
by `subjectAccessReviewVersion`.

When no matchConditions are satisfied for a request, the webhook would be skipped. In such
situations, the decision is logged in the audit log with the `webhookskipped.k8s.io` annotation.
Benefit of this is that resource and user info will also be logged.

The code path for enabling the above will only be triggered if the feature flag is enabled
while the feature is in alpha and beta.

### Monitoring

We will add the following 4 metrics:

1. `apiserver_authorization_decisions_total`

This will be incremented on round-trip of an authorizer. It will track total
authorization decision invocations across the following labels.

Labels {along with possible values}:
- `mode` {<authorizer_name>} # when authorizer is a webhook, prepend `webhook_`
- `decision` {Allow, Deny}

**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.

> If there is only one webhook and no name specified, there would be no `_<name>` suffix.

Reviewer: Currently the `name` field is only required when there are multiple webhooks. What this is saying is if there's only one webhook and the name is NOT specified, then there's no `_<name>` suffix just `webhook`.  But otherwise, whenever `name` is specified, it will be added as a suffix `webhook_<name>`.

> If there is only one webhook and no name specified, there would be no `_<name>` suffix.

Reviewer 5: I think it is far simpler to just make the name required.

> If there is only one webhook and no name specified, there would be no `_<name>` suffix.

Reviewer 7: > I think it is far simpler to just make the name required.

Yeah this was basically what I was saying.

> If there is only one webhook and no name specified, there would be no `_<name>` suffix.

Author: After thinking through this, made `name` a require field.

- Unit tests and Integration tests to be written

#### Future Alpha versions

- Revisit on the items mentioned in Non-Goals and see if any needs to be implemented

#### Beta

- Address user reviews and iterate (if needed, keep in Alpha until changes stabilize)
- Feature flag will be turned on by default

#### GA

- Feature flag removed
- Existing command line flags will be marked as deprecated and config file will take
precedence over the old flags

#### GA + 3 cycles

- Remove the existing command line flags

### Upgrade / Downgrade Strategy

While the feature is in Alpha, there is no change if cluster administrators want to
keep on using command line flags.

When the feature goes to Beta/GA or the cluster administrators want to configure
authorizers using a config file, they need to make sure the config file exists before
upgrading the cluster. Similarly when downgrading clusters, they would need to add
the flags back to their bootstrap mechanism.

### Version Skew Strategy

Not applicable.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate

>   - Feature gate name: `AuthorizationConfig`

Reviewer 5: ```suggestion
  - Feature gate name: `StructuredAuthorizationConfiguration`
```

        #   server.
        # - InClusterConfig: use the in-cluster configuration to call the
        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
        #   allowed for kube-apiserver.
        type: KubeConfig
        # Path to KubeConfigFile for connection info
        # Required, if connectionInfo.Type is KubeConfig
        kubeConfigFile: /kube-system-authz-webhook.yaml
        # matchConditions is a list of conditions that must be met for a request to be sent to this
        # webhook. An empty list of matchConditions matches all requests.
        # There are a maximum of 64 match conditions allowed.
        #
        # The exact matching logic is (in order):
        #   1. If at least one matchCondition evaluates to FALSE, then the webhook is skipped.
        #   2. If ALL matchConditions evaluate to TRUE, then the webhook is called.
        #   3. If at least one matchCondition evaluates to an error (but none are FALSE):
        #      - If failurePolicy=Deny, then the webhook rejects the request
        #      - If failurePolicy=NoOpinion, then the error is ignored and the webhook is skipped

>       matchConditions:

Reviewer 8: Sorry for the drive-by review, but a reasonable simplification might be to use CEL `&&` rather than multiple expressions since the semantics described here mirror those provided by CEL. @jpbetz

Reviewer 5: IMO the aesthetics of multiple expressions is much easier to read than `&&`ing a bunch of checks, especially if the checks are complex.

Reviewer 8: @enj the behavior, clarity, and safety should come before aesthetics. In the case, the cost is greater than the benefit. This proposal introduces a second implicit language into the authorization layer on top of CEL which complicates what users must reason about. At first glance, I assumed these expressions were OR'd since that's how many Cloud authorization policies behave. 

The following is strictly less clear than `a && b`

```
expression: a
expression: b
```

And maybe less desirable when considering what happens with the following:

```
expression: a || b && c
expression: d
```

Someone could accidentally assume that this means the same thing as:

```
expression: a || b
expression: c
expression: d
```

However, the correct expansion would be:

```
expression: a || b
expression: a || c
expression: d
```

In short, you'd added much more complexity and risk of an incorrect refactor without much value.

Reviewer 5: > At first glance, I assumed these expressions were OR'd since that's how many Cloud authorization policies behave.

We already use lists of CEL expressions in other Kube APIs, and they are AND'd together.  So for us, this is the most consistent approach.

Reviewer 5: Also the simple examples fail to show the clarity we are going for.  Multiple expressions are far easier to read:

```yaml
- expression: has(request.resourceAttributes)
- expression: request.resourceAttributes.namespace == 'my-ns'
- expression: request.resourceAttributes.name == 'my-name'
```

vs.

```yaml
expression: has(request.resourceAttributes) && request.resourceAttributes.namespace == 'my-ns' && request.resourceAttributes.name == 'my-name'
```

We are also not stopping anyone from using a single large expression if that is more clear to them.

Some other examples of lists of CEL expressions in Kube APIs:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-config-for-oidc-authentication#configuration-file

https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control#api-shape
